Source/DafnyCore/AST/MemberDecls.cs

```diff
@@ -467,6 +468,46 @@ public class Method : MemberDecl, TypeParameter.ParentType, IMethodCodeContext {
   public override bool IsOverrideThatAddsBody => base.IsOverrideThatAddsBody && Body != null;
   private static BlockStmt emptyBody = new BlockStmt(Token.NoToken, Token.NoToken, new List<Statement>());
 
+  public bool HasPostcondition =>
+    Ens.Count > 0 || Outs.Any(f => f.Type.AsSubsetType is not null);
+
+  public bool HasPrecondition =>
+    Req.Count > 0 || Ins.Any(f => f.Type.AsSubsetType is not null);
+
+  public override IEnumerable<AssumptionDescription> Assumptions() {
+    foreach (var a in base.Assumptions()) {
+      yield return a;
+    }
+
+    if (DeepChildren().Any(n => n is AssumeStmt)) {
+      yield return AssumptionDescription.AssumeInBody;
+    }
+
+    if (DeepChildren().Any(n => n is ForallStmt forallStmt && forallStmt.Body is null)) {
+      yield return AssumptionDescription.ForallWithoutBody;
+    }
+
+    if (DeepChildren().Any(n => n is OneBodyLoopStmt loopStmt && loopStmt.Body is null)) {
```

Comment:
Traversing `DeepChildren()` three times, while in practice it probably doesn't matter much, doesn't look great. I think changing this into a single `foreach` would be better.

Reply:
Even better would probably be to adopt your suggestion above to make this be an assumption on the loop itself, rather than on the method or function, and have member declarations yield all assumptions from their descendants. I'll make that change.